_posts/2025-04-09-phased-strimzi-upgrade-example.md

```diff
@@ -0,0 +1,733 @@
+              value: "strimzi-cluster-operator-0-43-0"
+```
+
+Contents of `install-files/cluster-operator-0.43.0/kustomization.yaml`
+```yaml
+resources:
+- 010-ServiceAccount-strimzi-cluster-operator.yaml
+...
+- 060-Deployment-strimzi-cluster-operator.yaml
+
+nameSuffix: '-0-43-0'
+
+patches:
+  - path: patches/patch-deployment.yaml
+    target:
+      kind: Deployment
+      name: strimzi-cluster-operator
+```
+
+##### Patching Role and ClusterRole Bindings
+
+The `RoleBinding` and `ClusterRoleBinding` resources used by the operator also need to be tailored for each Strimzi version.
+Again we can apply a patch to these in the `kustomization.yaml`:
+
+Contents of `install-files/cluster-operator-0.43.0/patches/patch-role-references.yaml`
+```yaml
+- op: replace
+  path: /subjects/0/name
+  value: strimzi-cluster-operator-0-43-0
+- op: replace
+  path: /subjects/0/namespace
+  value: strimzi
+```
+
+Contents of `install-files/cluster-operator-0.43.0/kustomization.yaml`
+```yaml
+resources:
+- 010-ServiceAccount-strimzi-cluster-operator.yaml
+...
+- 060-Deployment-strimzi-cluster-operator.yaml
+
+nameSuffix: '-0-43-0'
+
+patches:
+  ...
+  - path: patches/patch-role-references.yaml
+    target:
+      kind: (RoleBinding|ClusterRoleBinding)
+      name: ".*"
+```
+
+This patch updates all `RoleBinding` and `ClusterRoleBinding` resources in the resources list. 
+It replaces the name of the first entry in the subjects list (which is the operator’s `ServiceAccount` in the base installation files) with the name specified in the patch.
+This name should be the original name in the base Strimzi  operator `ServiceAccount` plus the `nameSuffix`.
+It will also patch the `namespace` which should match the namespace where the service account is installed. 
+In this case we assume all operator versions are installed in the `strimzi` namespace.
+
+##### Watching all namespaces
+
+The final thing we need to setup is configuring the namespaces that the Strimzi Operators will watch.
+For this example we assume that the various Strimzi versions will be watching all namespaces on the Kubernetes cluster.
+As per [the documentation](https://strimzi.io/docs/operators/latest/deploying#deploying-cluster-operator-to-watch-whole-cluster-str) we need to alter the `RoleBinding` and `ClusterRoleBinding` namespaces to match the one the operator is installed in.
+However, that is already handled by the step above.
+
+We then need to set the `STRIMZI_NAMESPACE` env var in the operator `Deployment` to `*`.
+For this we can update the `patch-deployment.yaml` from the earlier step:
+
+Contents of `install-files/cluster-operator-0.43.0/patches/patch-deployment.yaml`
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: strimzi-cluster-operator
+spec:
+  template:
+    spec:
+      containers:
+        - name: strimzi-cluster-operator
+          env:
+            - name: STRIMZI_NAMESPACE
+              value: "*" 
+              valueFrom: null
```

Comment:
Is the `valueFrom: null` intentional here? I assume it is valid Deployment this way ... but it looks strange and would be normally omitted.

Reply:
So this is required in order to negate the existing `valueFrom` in the base resource the patch is applied to. The resulting output would not contain the `valueFrom` field.